src/components/tl/ucp/alltoall/alltoall_bruck.c

```diff
@@ -0,0 +1,240 @@
+        ucc_assert(send_buffer_index >= 0);
+        snd_offset = send_buffer_index * seg_size;
+        st = ucc_mc_memcpy(PTR_OFFSET(dst, seg_size *
+                                      ((trank - index + tsize) % tsize)),
+                           PTR_OFFSET(src, snd_offset), seg_size,
+                           UCC_MEMORY_TYPE_HOST, UCC_MEMORY_TYPE_HOST);
+        if (ucc_unlikely(st != UCC_OK)) {
+            return st;
+        }
+    }
+
+    return UCC_OK;
+}
+
+ucc_status_t ucc_tl_ucp_alltoall_bruck_finalize(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_ucp_task_t *task = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_status_t st, global_st;
+
+    global_st = ucc_mc_free(task->alltoall_bruck.scratch_mc_header);
+    if (ucc_unlikely(global_st != UCC_OK)) {
+        tl_error(UCC_TASK_LIB(task), "failed to free scratch buffer");
+    }
+
+    st = ucc_tl_ucp_coll_finalize(&task->super);
+    if (ucc_unlikely(st != UCC_OK)) {
+        tl_error(UCC_TASK_LIB(task), "failed finalize collective");
+        global_st = st;
+    }
+    return global_st;
+}
+
+void ucc_tl_ucp_alltoall_bruck_progress(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_ucp_task_t *task       = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_tl_ucp_team_t *team       = TASK_TEAM(task);
+    ucc_rank_t         trank      = UCC_TL_TEAM_RANK(team);
+    ucc_rank_t         tsize      = UCC_TL_TEAM_SIZE(team);
+    ucc_coll_args_t   *args       = &TASK_ARGS(task);
+    void              *scratch    = task->alltoall_bruck.scratch_mc_header->addr;
+    void              *mergebuf   = args->dst.info.buffer;
+    const ucc_rank_t   nrecv_segs = tsize / 2;
+    const size_t       seg_size   = ucc_dt_size(args->src.info.datatype) *
+                                    args->src.info.count / tsize;
+    void              *data;
+    ucc_rank_t sendto, recvfrom, step, index;
+    ucc_rank_t level, snd_count;
+    int send_buffer_index;
+    ucc_status_t st;
+
+    if (task->alltoall_bruck.phase == PHASE_SENDRECV) {
+        goto ALLTOALL_BRUCK_PHASE_SENDRECV;
+    }
+
+    step = 1 << (task->alltoall_bruck.iteration - 1);
+    while (step < tsize) {
+        level = task->alltoall_bruck.iteration - 1;
+        sendto = get_bruck_send_peer(trank, tsize, step, 1);
+        recvfrom = get_bruck_recv_peer(trank, tsize, step, 1);
+
+        snd_count = 0;
+        for (index = get_bruck_step_start(step, 1);
+             index <= get_bruck_step_finish(tsize - 1, RADIX, 1, step);
+             index = GET_NEXT_BRUCK_NUM(index, RADIX, step)) {
+            send_buffer_index = find_seg_index(index, level + 1, nrecv_segs);
+            if (send_buffer_index == -1) {
+                data = PTR_OFFSET(args->src.info.buffer,
+                                  ((index + trank) % tsize) * seg_size);
+            } else {
+                data = PTR_OFFSET(scratch, send_buffer_index * seg_size);
+            }
+            ucc_mc_memcpy(PTR_OFFSET(mergebuf, seg_size * snd_count),
```

Comment:
@Sergei-Lebedev using UCP datatype and ucp_dt_make_contig would be any better here?

Reply:
yes, i tried that, we can save extra memcpy but it only helps at larger msg. But for these messages pairwise is better anyway.